File: LIB/2024.01.25/Q5014_스타트링크/Q5014.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Q5014 {

  public static int F, S, G, U, D;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader((System.in)));
    StringTokenizer st = new StringTokenizer(br.readLine());

    F = Integer.parseInt(st.nextToken());
    S = Integer.parseInt(st.nextToken());
    G = Integer.parseInt(st.nextToken());
    U = Integer.parseInt(st.nextToken());
    D = Integer.parseInt(st.nextToken());

    bfs();
  }

  public static void bfs() {
    Queue<Point> queue = new LinkedList<>();
    boolean[] visited = new boolean[F + 1];

    queue.add(new Point(S, 0));
    visited[S] = true;

    while(!queue.isEmpty()) {
      Point cur = queue.poll();

      if(cur.floor == G) {
        System.out.println(cur.dist);
        return;
      }

      int nu = cur.floor + U;
      int nd = cur.floor - D;

      if(nu <= F && !visited[nu]) {
        queue.add(new Point(nu, cur.dist + 1));
        visited[nu] = true;
      }

      if(0 < nd && !visited[nd]) {
        queue.add(new Point(nd, cur.dist + 1));
        visited[nd] = true;
      }
    }

    System.out.println("use the stairs");
  }

  static class Point {
    int floor;
    int dist;
  
    Point(int f, int d) {
      this.floor = f;
      this.dist = d;
    }
  }
}

File: LIB/2024.01.25/Q5014_스타트링크/Q5014.py
# import sys
# from collections import deque
# F, S, G, U, D = map(int, sys.stdin.readline().strip().split())
# move_cnt = 0

# def bfs():
#   visited = [False] * (F + 1)
#   queue = deque()

#   queue.append((S, 0))
#   visited[S] = True

#   while queue:
#     cur, dist = queue.popleft()

#     if cur == G:
#       print(dist)
#       return

#     next_up = cur + U
#     next_down = cur - D

#     if next_up <= F and not visited[next_up]:
#       visited[next_up] = True
#       queue.append((next_up, dist + 1))

#     if 0 < next_down and not visited[next_down]:
#       visited[next_down] = True
#       queue.append((next_down, dist + 1))

#   print("use the stairs")

# bfs()

#####

import sys
from collections import deque
F, S, G, U, D = map(int, sys.stdin.readline().strip().split())
move_cnt = 0

def bfs():
  visited = [-1] * (F + 1)
  queue = deque()

  queue.append(S)
  visited[S] = 0

  while queue:
    cur = queue.popleft()

    if cur == G:
      print(visited[cur])
      return
    
    for next_floor in [(cur + U), (cur - D)]:
      if 0 < next_floor <= F and visited[next_floor] == -1:
        visited[next_floor] = visited[cur] + 1
        queue.append(next_floor)

  print("use the stairs")

bfs()